test/prop_thoas.erl

```diff
@@ -0,0 +1,236 @@
+       begin
+           Encoded = thoas:encode(Input, EncOpts),
+           ?assertMatch({ok, _}, thoas:decode(Encoded)),
+           true
+       end).
+
+prop_truncated_no_crash() ->
+    ?FORALL(
+       {Json, CutAt},                         % at least 2 char JSON and position in it where to cut
+       ?LET(Json,
+            ?SUCHTHAT(Json,
+                      ?LET({Term, EncOpts},
+                           {json_term(), encode_options()},
+                           thoas:encode(Term, EncOpts)),
+                      byte_size(Json) > 1),
+            {Json, integer(1, byte_size(Json) - 1)}),
+       begin
+           Part = binary:part(Json, 0, CutAt),
+           case thoas:decode(Part) of
+               {error, unexpected_end_of_input} ->
+                   true;
+               {error, {unexpected_byte, _, _}} ->
+                   true;
+               {ok, Num} when is_number(Num) ->
+                   %% Integers can be truncated without anyone noticing
+                   true;
+               Other ->
+                   error({unexpected_return, Other})
+           end
+       end).
+
+prop_thoas_encode_jsx_decode_match() ->
+    ?FORALL(
+       Input,
+       json_term(),
+       begin
+           Encoded = thoas:encode(Input),
+           ?assertEqual(Input, jsx:decode(Encoded)),
+           true
+       end).
+
+prop_jsx_encode_thoas_decode_match() ->
+    ?FORALL(
+       Input,
+       json_term(),
+       begin
+           Encoded = jsx:encode(Input),
+           ?assertEqual({ok, Input}, thoas:decode(Encoded)),
+           true
+       end).
+
+%% Not realy reliable
+%% prop_encode_not_slower_than_jsx() ->
+%%     ?FORALL(
+%%        Input,
+%%        ?SIZED(Size, json_term(Size + 10)),
+%%        begin
+%%            erlang:garbage_collect(),
+%%            {JsxTime, _} = timer:tc(run_n_f(fun() -> jsx:encode(Input) end, 50)),
+%%            erlang:garbage_collect(),
+%%            {ThoasTime, _} = timer:tc(run_n_f(fun() -> thoas:encode(Input) end, 50)),
+%%            ?assert(ThoasTime / JsxTime < 2,
+%%                    [{thoas_time, ThoasTime},
+%%                     {jsx_time, JsxTime}]),
+%%            true
+%%        end).
+
+%% prop_decode_not_slower_than_jsx() ->
+%%     ?FORALL(
+%%        Input,
+%%        ?SIZED(Size, json_term(Size + 10)),
+%%        begin
+%%            JsonBin = thoas:encode(Input),
+%%            erlang:garbage_collect(),
+%%            {JsxTime, _} = timer:tc(run_n_f(fun() -> jsx:decode(JsonBin) end, 50)),
+%%            erlang:garbage_collect(),
+%%            {ThoasTime, _} = timer:tc(run_n_f(fun() -> thoas:decode(JsonBin) end, 50)),
+%%            ?assert(ThoasTime / JsxTime < 2,
+%%                    [{thoas_time, ThoasTime},
+%%                     {jsx_time, JsxTime}]),
+%%            true
+%%        end).
+
+%% run_n_f(Fun, N) ->
+%%     fun() -> run_n(Fun, N) end.
+
+%% run_n(_Fun, 0) -> ok;
+%% run_n(Fun, N) -> 
+%%     Fun(),
+%%     run_n(Fun, N - 1).
+
```

Comment:
```suggestion
```

Reply:
Sure. I tried to use it to try to find some potential pathological corner-cases where `thoas` would run significantly slower than jsx on some weird input. But those tests were not really reliable with condition "fail if thoas is 2x slower than jsx on this input term" to run them on CI on a regular base. Maybe I can change the condition to "fail if thoas is 10x slower" or somehow exclude them from CI?
But can remove as well, no problem.

Reply:
I like the idea but I think it would need to be done with a proper benchmarking framework and not run on CI. From experience there's too much variance to get reliable results otherwise.